src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs

```diff
@@ -34,8 +34,46 @@ public CertificateCallbackMapper(Func<HttpRequestMessage, X509Certificate2?, X50
             }
         }
 
-        public static async ValueTask<Stream> ConnectAsync(string host, int port, CancellationToken cancellationToken)
+        public static async ValueTask<Stream> ConnectAsync(string host, int port, bool async, CancellationToken cancellationToken)
         {
+            // For synchronous connections, we can just create a socket and make the connection.
+            // (For async, we need to do more gymnastics with SocketAsyncEventArgs.)
+            if (!async)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                Socket? socket = new Socket(SocketType.Stream, ProtocolType.Tcp);
+                try
+                {
+                    socket.NoDelay = true;
+                    using (cancellationToken.UnsafeRegister(s => ((Socket)s!).Dispose(), socket))
+                    {
+                        socket.Connect(new DnsEndPoint(host, port));
+                    }
+                }
+                catch (Exception e)
+                {
+                    socket.Dispose();
+
+                    if (CancellationHelper.ShouldWrapInOperationCanceledException(e, cancellationToken))
+                    {
+                        // Cancellation was requested, so assume that the failure is due to
+                        // the cancellation request. This is a bit unorthodox, as usually we'd
+                        // prioritize a non-OperationCanceledException over a cancellation
+                        // request to avoid losing potentially pertinent information.  But given
+                        // the cancellation design where we tear down the underlying stream upon
+                        // a cancellation request, which can then result in a myriad of different
+                        // exceptions (argument exceptions, object disposed exceptions, socket exceptions,
+                        // etc.), as a middle ground we treat it as cancellation, but still propagate the
+                        // original information as the inner exception, for diagnostic purposes.
+                        throw CancellationHelper.CreateOperationCanceledException(e, cancellationToken);
+                    }
+
+                    throw;
+                }
+
+                return new NetworkStream(socket, ownsSocket: true);
```

Comment:
Nit: can we move the body of this if block into a separate Connect method that's just called from here?  I don't know how many call sites ConnectAsync has, but if it's just one, it'd be even better for the if to be done at the call site, branching between Connect and ConnectAsync.

Reply:
Two occurrences: 
https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L1084
https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L1088
Should I also move the async block into its own method and have this method be simple:
```C#
return async ? ConnectAsync(...) : new ValueTask<Stream>(Connect(...));
```

Reply:
Sure.